### PR TITLE
[JENKINS-46124] - support map with EnvStep

### PR DESCRIPTION
**Pull request by jetersen**
fixes [JENKINS-46124](https://issues.jenkins-ci.org/browse/JENKINS-46124)

### CODE REVIEW

File: src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java [hunk 1]
```diff
@@ -83,6 +83,36 @@
...
+                story.j.assertLogContains(Functions.isWindows() ? "inside CUSTOM=override NOVEL=val BUILD_TAG=custom NULLED= FOOPATH=C:\\ball;C:\\foos;" : "inside CUSTOM=override NOVEL=val BUILD_TAG=custom NULLED= FOOPATH=/opt/ball:/opt/foos:", b);
+                story.j.assertLogContains("groovy NULLED=null", b);
+                story.j.assertLogContains("outside CUSTOM=initial NOVEL= NULLED=outside", b);
+                List<FlowNode> coreStepNodes = new DepthFirstScanner().filteredNodes(
+                    b.getExecution(),
+                    Predicates.and(
+                        new NodeStepTypePredicate("withEnv"),
+                        n -> n instanceof StepStartNode && !((StepStartNode) n).isBody()));
+                assertThat(coreStepNodes, Matchers.hasSize(1));
+                assertEquals("CUSTOM, NOVEL, BUILD_TAG, NULLED, FOOPATH+BALL", ArgumentsAction.getStepArgumentsAsString(coreStepNodes.get(0)));
```
Comment (casz) [inline src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java]: This test case fails, I am not sure why it seems to be doing the right thing but unable to get the step arguments.

Comment (casz) [reply src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java]: d0h I had to write the converter for arguments to string.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java [hunk 2]
```diff
@@ -150,11 +153,36 @@
...
             } else {
                 return null;
             }
         }
 
+        @NonNull
+        @Override
+        public Map<String, Object> customInstantiate(@NonNull Map<String, Object> arguments) {
+            Map<String, Object> r = new HashMap<>(arguments);
+            Object overrides = r.get("overrides");
```
Comment (MindaugasLaganeckas) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: DRY for "overrides" :)

**Review by MindaugasLaganeckas — APPROVED** [review]
A nice feature and a nice set of tests! :)

File: src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java [hunk 3]
```diff
@@ -150,11 +153,37 @@
                     }
                 }
                 return b.toString();
+            } else if (overrides instanceof Map) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: BTW I lobbied for `ArgumentsAction[Impl]` to persist only the internal, resolved form of arguments, but the design we went with persists the surface form, necessitating extra work like this.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java [hunk 4]
```diff
@@ -150,11 +153,37 @@
...
+            Object overrides = r.get(key);
+            if (overrides instanceof Map) {
+                r.put(key, toKeyValueList((Map<?, ?>) overrides));
+            }
+            return r;
+        }
+
+        private static List<String> toKeyValueList(Map<?, ?> map) {
+            return map.entrySet().stream()
+                .filter(e -> e.getKey() instanceof String)
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: Clearer IMO to just cast to `String`, so that people will get a clear error if they use something else.
```suggestion
```

File: src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java [hunk 5]
```diff
@@ -150,11 +153,37 @@
...
+                .collect(Collectors.toMap(e -> (String)e.getKey(), e -> e.getValue() == null ? "" : e.getValue().toString()))
+                .entrySet().stream()
+                .map(m -> m.getKey() + "=" + m.getValue())
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: Why are you collecting a map and then restreaming it? Would seem to simpler to just

```java
.map(m -> (String) e.getKey() + "=" + (m.getValue() == null ? "" : m.getValue()))

File: src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java [hunk 6]
```diff
@@ -83,6 +83,36 @@
         });
     }
 
+    @Test public void mapOverriding() {
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                p.setDefinition(new CpsFlowDefinition(
+                    "env.CUSTOM = 'initial'\n" +
+                        "env.FOOPATH = node {isUnix() ? '/opt/foos' : 'C:\\\\foos'}\n" +
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java]: You need not copy all these exotic corner cases from overriding. Should suffice to have a simple script like

```groovy
withEnv([a: 1, b: 2]) {
  echo "a=$a b=$b"
}
```

File: src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java [hunk 7]
```diff
@@ -105,6 +135,28 @@
         });
     }
 
+    @Test public void mapParallel() {
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java]: Again, unnecessary, adequately covered by existing tests; all that is changing is the syntax.

File: src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java [hunk 8]
```diff
@@ -140,6 +192,41 @@
         });
     }
 
+    @Test public void mapRestarting() {
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java]: Ditto.

File: src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java [hunk 9]
```diff
@@ -158,6 +245,44 @@
         });
     }
 
+    @Test public void mapNumbersNested() {
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java]: Some simpler variant of this may still be useful, to assert that map syntax allows non-string values.

File: src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java [hunk 10]
```diff
@@ -158,6 +245,44 @@
...
+                        "    }\n" +
+                        "  }\n" +
+                        "}", true));
+                WorkflowRun b = story.j.assertBuildStatusSuccess(p.scheduleBuild2(0));
+                story.j.assertLogContains("A=1 B=2 C=true", b);
+            }
+        });
+    }
+
+    @Test public void mapNested() {
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java]: Probably can be skipped.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java [hunk 11]
```diff
@@ -150,11 +153,37 @@
                     }
                 }
                 return b.toString();
+            } else if (overrides instanceof Map) {
+                return ((Map<?, ?>) overrides).keySet()
+                    .stream()
+                    .filter(e -> e instanceof String)
+                    .map(Object::toString)
+                    .collect(Collectors.joining(", "));
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: Not covered by tests AFAICT?

File: src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java [hunk 12]
```diff
@@ -150,11 +153,37 @@
...
             } else {
                 return null;
             }
         }
 
+        @NonNull
+        @Override
+        public Map<String, Object> customInstantiate(@NonNull Map<String, Object> arguments) {
+            Map<String, Object> r = new HashMap<>(arguments);
+            final String key = "overrides";
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: This design forces people to write, say,

```groovy
withEnv([VAR: 'value']) {/* … */}
```

rather than the possibly more natural

```groovy
withEnv(VAR: 'value') {/* … */}
```

which would be available for anything other than an environment variable named (weirdly) `overrides`.

**Review by jglick — COMMENTED** [review]
Do you want to implement `customUninstantiate` to guide users of **Pipeline Syntax** to use the new form, or leave them with the `List<String>` syntax?

File: src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java (hunk 11 above)
Comment (casz) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: Definitely covered by tests as it was failing before I created the converter 😆

Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: Meaning tests verify that this method returns normally when a map is used, not that the return value is correct.

Comment (casz) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: No there is a test that looks at the output.
https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/105/files#diff-c2e520244ad09761e2187513ef5c751cR111

File: src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java (hunk 12 above)
Comment (casz) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: Hmmm I think we need to support both as we are trying to support passing a map from scm checkout

Comment (casz) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: it's named overrides because that's how groovy passes the object.
image

Comment (casz) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: Even when I pass it like you suggested
```java
withEnv(VAR: 'value') {/* … */}
```
I still get `arguments.overrides` with whatever arguments inside.

Comment (casz) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: I'll add a test case for this

File: src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java (hunk 11 above)
Comment (casz) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: To the two tests I added now I am testing the output of the function.

**Comment by casz** [comment]
please take another look @jglick I think I listen to your feedback, I honestly don't know what I was thinking about when I wrote that stream collector 🤯

File: src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java [hunk 13]
```diff
@@ -245,40 +158,48 @@
...
                 WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
                 p.setDefinition(new CpsFlowDefinition(
                     "node {\n" +
-                        "  withEnv([A: 1]) {\n" +
-                        "    withEnv([B: 2]) {\n" +
-                        "      withEnv([C: true]) {\n" +
-                        "        isUnix() ? sh('echo A=$A B=$B C=$C') : bat('echo A=%A% B=%B% C=%C%')\n" +
-                        "      }\n" +
-                        "    }\n" +
+                        "  withEnv(a: 1, b: 2, c: 'hello world', d: true) {\n" +
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java]: test null values

File: src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java [hunk 14]
```diff
@@ -178,10 +178,7 @@
 
         private static List<String> toKeyValueList(Map<?, ?> map) {
             return map.entrySet().stream()
-                .filter(e -> e.getKey() instanceof String)
-                .collect(Collectors.toMap(e -> (String)e.getKey(), e -> e.getValue() == null ? "" : e.getValue().toString()))
-                .entrySet().stream()
-                .map(m -> m.getKey() + "=" + m.getValue())
+                .map(m -> m.getKey() + "=" + (m.getValue() == null ? "" : m.getValue().toString()))
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]:
```suggestion
                .map(m -> (String) m.getKey() + "=" + (m.getValue() == null ? "" : m.getValue().toString()))
```

since non-`String` keys would be bogus

File: src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java [hunk 15]
```diff
@@ -245,40 +158,48 @@
...
+                        "    echo \"a=$a b=$b c=$c d=$d\"" +
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java]: arguably more legibly

```suggestion
                        "    echo(/a=$a b=$b c=$c d=$d/)" +
```

**Review by jglick — APPROVED** [review]

File: src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java (hunk 14 above)
Comment (casz) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: IDE consider this a redundant cast and wants to remove it.

File: src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java (hunk 13 above)
Comment (casz) [reply src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java]: the customInstate does the right with with `e: null` turning into `e=` however it throws a `groovy.lang.MissingPropertyException: No such property: e for class: groovy.lang.Binding`

Comment (casz) [reply src/test/java/org/jenkinsci/plugins/workflow/steps/EnvStepTest.java]: ah when null we have to echo like this `E=${env.E}`

File: src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java (hunk 14 above)
Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: Then your IDE is wrong and you should ignore it. You should be able to have a test demonstrating a failure from some bogus thing like

```groovy
withEnv([new Object(): 'x']) {}
```

Comment (casz) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: This case is not supported:
```
WorkflowScript: 2: illegal colon after argument expression;
  solution: a complex label expression before a colon must be parenthesized @ line 2, column 72.
  D: true, E: null, new Object(): 'x']) {
```

Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: Maybe

```groovy
withEnv([(new Object()): 'x']) {}
```

then, whatever the syntax is. At worst

```groovy
def m = [:]
m.put(new Object(), 'x')
withEnv(m) {}
```

Comment (casz) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: Than you get into the whole script security: one you add `(new Object())` `org.jenkinsci.plugins.scriptsecurity.sandbox.RejectedAccessException: Scripts not permitted to use new java.lang.Object`

Comment (casz) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: I'd argue your asking for trouble

Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: OK forget `new Object()`, just some arbitrary object key which is _not_ a `String`. `new URL('http://nowhere.net/')` for example is whitelisted.

Comment (casz) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: Than you get into `java.lang.ClassCastException: java.net.URL cannot be cast to java.lang.String` 🤔 

Properly better to call `toString` instead of trying to cast the object

Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: > Then you get into `java.lang.ClassCastException`

Exactly what I was requesting with 8d385a2.

Comment (casz) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: image

**Comment by casz** [comment]
@bitwiseman @jglick 🙏 Would be a great addition to get merged.

**Review by dwnusbaum — CHANGES_REQUESTED** [review]
Thanks for the PR! The implementation looks good to me (thanks for adding tests!). I am a little hesitant to add complexity since we will have to support both syntaxes going forward, but I think the trade-off is worth it in this case.

Can you please update [EnvStep/help-overrides.html](https://github.com/jenkinsci/workflow-basic-steps-plugin/blob/f5586e7f38df6ab42688c3c968943874d418461a/src/main/resources/org/jenkinsci/plugins/workflow/steps/EnvStep/help-overrides.html) (and maybe also [EnvStep/help.html](https://github.com/jenkinsci/workflow-basic-steps-plugin/blob/f5586e7f38df6ab42688c3c968943874d418461a/src/main/resources/org/jenkinsci/plugins/workflow/steps/EnvStep/help.html)) to mention the new syntax and give an example? Once that is done I am happy to merge the PR.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java (hunk 3 above)
Comment (bitwiseman) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: @jglick  Okay, is there some action to be taken about that now?

Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java]: Not easily. Just noting some background.

**Comment by casz** [comment]
I'll see what I can do :)
However somewhat busy in this week. Starting at a new workplace 😁

**Comment by mss** [comment]
@dwnusbaum I was looking for this exact feature for years and finally found this PR but it got stalled. I'd be happy to provide the requested documentation but since this code does not apply in the meantime I wondered who would have to adapt it to the changed codebase?
